src/py/flwr/simulation/ray_transport/ray_actor.py

```diff
@@ -0,0 +1,280 @@
+
+    def terminate(self):
+        """Terminate Actor."""
+        log(WARNING, f"Manually terminating {self.__class__.__name__}")
+        ray.actor.exit_actor()
+
+    def run(self, client_fn: Callable, client_id):
+        """Run a client workload."""
+        # execute tasks and return result
+        # return also cid which is needed to ensure results
+        # from the pool are correctly assigned to each ClientProxy
+        try:
+            client_results = client_fn()
+        except Exception as ex:
+            client_trace = traceback.format_exc()
+            message = (
+                "\n\tSomething went wrong when running your client workload."
+                f"\n\tClient {client_id} crashed when the {self.__class__.__name__}"
+                " was running its workload."
+                f"\n\tException triggered on the client side: {client_trace}"
+            )
+            raise ClientException(message) from ex
+
+        return client_id, client_results
+
+
+def pool_size_from_resources(client_resources: Dict):
+    """Calculate number of Actors that fit in pool given the resources in the.
+
+    cluster and those required per client.
+    """
+    cluster_resources = ray.cluster_resources()
+    num_cpus = cluster_resources["CPU"]
+    num_gpus = cluster_resources.get("GPU", 0)  # there might not be GPU
+    num_actors = int(num_cpus / client_resources["num_cpus"])
+    # if a GPU is present and client resources do require one
+    if client_resources["num_gpus"] > 0.0:
+        if num_gpus:
+            # if there are gpus in the cluster
+            num_actors = min(num_actors, int(num_gpus / client_resources["num_gpus"]))
+        else:
+            num_actors = 0
+
+    if num_actors == 0:
+        log(
+            WARNING,
+            f"Your ActorPool is empty. Your system ({num_cpus = }, {num_gpus = }) "
+            "does not meet the criteria to host at least one client with resources:"
+            f" {client_resources}. Consider lowering your `client_resources`",
+        )
+
+    return num_actors
+
+
+class VirtualClientEngineActorPool(ActorPool):
+    """A pool of VirtualClientEngine Actors."""
+
+    def __init__(self, client_resources: Dict, max_restarts: int = 1):
+        self.client_resources = client_resources
+        self.actor_max_restarts = max_restarts
+        num_actors = pool_size_from_resources(client_resources)
+        actors = [
+            VirtualClientEngineActor.options(
+                **client_resources, max_restarts=max_restarts
+            ).remote()
+            for _ in range(num_actors)
+        ]
+
+        super().__init__(actors)
+
+        self._cid_to_future = {}  # a dict
+        self.actor_to_remove: Set[str] = set()  # a set
+        self.num_actors = len(actors)
+
+        self.lock = threading.RLock()
+
+        # TODO: asyncio check every N seconds if cluster has grown
+        # --> add more actors to the pool if so
```

Comment:
We could use such a mechanism also to replace the `max_restarts` setting. The cluster could also shrink, so monitoring the current size could be useful to either add or remove actors from the pool.

Reply:
yeah, the shrinking is already supported. An actor is added back to the pool after completing a job iif it fits back into the pool (see `process_unordered_future()` and in particular the logic surounding `_check_actor_fits_in_pool()`)